src/hotspot/share/gc/parallel/mutableSpace.cpp

```diff
@@ -235,7 +235,10 @@ void MutableSpace::oop_iterate(OopIterateClosure* cl) {
 void MutableSpace::object_iterate(ObjectClosure* cl) {
   HeapWord* p = bottom();
   while (p < top()) {
-    cl->do_object(cast_to_oop(p));
+    oop obj = cast_to_oop(p);
+    if (!obj->is_forwarded()) {
```

Comment:
Maybe some comment along these lines "When promotion-failure occurs during Young GC, eden/from space is not cleared, so we can encounter objs with "forwarded" markword. They are essentially dead, so skipping them."